File: unity-test/Assets/BleHidTest/Events/DeviceDisconnectedEvent.cs
using System;

namespace BleHid.Events
{
    /// <summary>
    /// Event fired when a device disconnects.
    /// </summary>
    public class DeviceDisconnectedEvent : DeviceEvent
    {
        /// <summary>
        /// The reason for the disconnection
        /// </summary>
        public DisconnectReason Reason { get; private set; }
        
        /// <summary>
        /// Creates a new device disconnected event.
        /// </summary>
        /// <param name="deviceAddress">MAC address of the device</param>
        /// <param name="reason">The reason for disconnection</param>
        /// <param name="eventId">ID of the event</param>
        /// <param name="javaTimestamp">Java timestamp (milliseconds since epoch)</param>
        public DeviceDisconnectedEvent(string deviceAddress, DisconnectReason reason, Guid eventId, long javaTimestamp)
            : base(BleHidEventType.DeviceDisconnected, deviceAddress, null, eventId, javaTimestamp)
        {
            Reason = reason;
        }
        
        /// <summary>
        /// Creates a new device disconnected event.
        /// For events created on the Unity side.
        /// </summary>
        /// <param name="deviceAddress">MAC address of the device</param>
        /// <param name="reason">The reason for disconnection</param>
        public DeviceDisconnectedEvent(string deviceAddress, DisconnectReason reason)
            : base(BleHidEventType.DeviceDisconnected, deviceAddress)
        {
            Reason = reason;
        }
        
        /// <summary>
        /// Gets a string representation of this event.
        /// </summary>
        public override string ToString()
        {
            return $"{base.ToString()}, reason={Reason}";
        }
    }
}

File: unity-plugin/src/main/java/com/example/blehid/unity/events/DeviceDisconnectedEvent.java
package com.example.blehid.unity.events;

import android.bluetooth.BluetoothDevice;

/**
 * Event fired when a device disconnects.
 */
public class DeviceDisconnectedEvent extends DeviceEvent {
    private final DisconnectReason reason;
    
    /**
     * Reasons why a device might disconnect
     */
    public enum DisconnectReason {
        NORMAL_DISCONNECT,
        CONNECTION_TIMEOUT,
        CONNECTION_LOST,
        REMOTE_DEVICE_TERMINATED,
        LOCAL_HOST_TERMINATED,
        UNKNOWN
    }
    
    /**
     * Creates a new device disconnected event.
     * 
     * @param device The disconnected device
     * @param reason The reason for disconnection
     */
    public DeviceDisconnectedEvent(BluetoothDevice device, DisconnectReason reason) {
        super(EventType.DEVICE_DISCONNECTED, device);
        this.reason = reason;
    }
    
    /**
     * Creates a new device disconnected event using just the address.
     * 
     * @param deviceAddress MAC address of the disconnected device
     * @param reason The reason for disconnection
     */
    public DeviceDisconnectedEvent(String deviceAddress, DisconnectReason reason) {
        super(EventType.DEVICE_DISCONNECTED, deviceAddress);
        this.reason = reason;
    }
    
    /**
     * Gets the reason for disconnection.
     * 
     * @return Disconnect reason
     */
    public DisconnectReason getReason() {
        return reason;
    }
    
    @Override
    public String toString() {
        return String.format("%s[reason=%s]", super.toString(), reason);
    }
}
